core/middlewares/with-analytics-cookies.ts

```diff
@@ -0,0 +1,24 @@
+import { NextRequest } from 'next/server';
+
+import { setVisitIdCookie, setVisitorIdCookie } from '~/lib/analytics/bigcommerce';
+
+import { MiddlewareFactory } from './compose-middlewares';
+
+export const withAnalyticsCookies: MiddlewareFactory = (next) => {
+  return async (request, event) => {
+    const visitorId = await setVisitorIdCookie();
+    const { id: visitId, isNew: isVisitNew } = await setVisitIdCookie();
+
+    if (isVisitNew) {
+      event.waitUntil(sendShopperVisitEvent(request, visitorId, visitId));
+    }
+
+    return next(request, event);
+  };
+};
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+async function sendShopperVisitEvent(request: NextRequest, visitorId: string, visitId: string) {
+  // fetch from request user agent, url and referer url
+  // TODO: Send shopper visit event to BigCommerce
+}
```

Comment:
Alternatively can [after](https://nextjs.org/docs/app/api-reference/functions/after) method be used for sending analytic events in non-blocking way, instead of middleware?

Reply:
We will consider this when will be implementing GQL request

Reply:
`waitUntil` has more deployment provider support than `after` does, so something to think about here.